File: apollo/src/main/java/io/muun/apollo/data/preferences/TransactionSizeRepository.java
package io.muun.apollo.data.preferences;


import io.muun.apollo.data.serialization.SerializationUtils;
import io.muun.apollo.domain.model.NextTransactionSize;
import io.muun.common.utils.Preconditions;

import android.content.Context;
import com.f2prateek.rx.preferences.Preference;
import rx.Observable;

import javax.annotation.Nullable;
import javax.inject.Inject;

public class TransactionSizeRepository extends BaseRepository {

    private static final String KEY_TRANSACTION_SIZE = "transaction_size";

    private final Preference<String> transactionSizePreference;

    @Inject
    public TransactionSizeRepository(Context context) {
        super(context);
        transactionSizePreference = rxSharedPreferences.getString(KEY_TRANSACTION_SIZE, null);
    }

    @Override
    protected String getFileName() {
        return "transaction_size";
    }

    /**
     * Get the stored NextTransactionSize.
     */
    @Nullable
    public NextTransactionSize getNextTransactionSize() {
        return watchNextTransactionSize().toBlocking().first();
    }

    /**
     * Watch the stored NextTransactionSize.
     */
    public Observable<NextTransactionSize> watchNextTransactionSize() {
        return transactionSizePreference.asObservable()
                .map(json -> {
                    if (json != null) {
                        return SerializationUtils.deserializeJson(NextTransactionSize.class, json);
                    } else {
                        return null;
                    }
                });
    }

    /**
     * Replace the stored NextTransactionSize.
     */
    public void setTransactionSize(NextTransactionSize transactionSize) {
        transactionSizePreference.set(
                SerializationUtils.serializeJson(NextTransactionSize.class, transactionSize)
        );
    }

    /**
     * Migration to init expected debt for pre-existing NTSs.
     */
    public void initExpectedDebt() {
        final NextTransactionSize nts = getNextTransactionSize();

        Preconditions.checkNotNull(nts);

        setTransactionSize(nts.initExpectedDebt());
    }
}

File: android/apollo/src/main/java/io/muun/apollo/data/preferences/TransactionSizeRepository.kt
package io.muun.apollo.data.preferences

import android.content.Context
import io.muun.apollo.data.preferences.rx.Preference
import io.muun.apollo.data.serialization.SerializationUtils
import io.muun.apollo.domain.model.NextTransactionSize
import io.muun.common.utils.Preconditions
import rx.Observable
import javax.inject.Inject

class TransactionSizeRepository @Inject constructor(
    context: Context,
    repositoryRegistry: RepositoryRegistry,
) : BaseRepository(context, repositoryRegistry) {

    companion object {
        private const val KEY_TRANSACTION_SIZE = "transaction_size"
    }

    private val transactionSizePreference: Preference<String>
        get() = rxSharedPreferences.getString(KEY_TRANSACTION_SIZE, null)

    override val fileName get() = "transaction_size"

    /**
     * Get the stored NextTransactionSize.
     */
    val nextTransactionSize: NextTransactionSize?
        get() = watchNextTransactionSize().toBlocking().first()

    /**
     * Watch the stored NextTransactionSize.
     * Only to be used after transactionSizePreference.isSet == true.
     */
    fun watchNonNullNts(): Observable<NextTransactionSize> {
        return watchNextTransactionSize().map { nts -> nts!! }
    }

    /**
     * Watch the stored NextTransactionSize.
     */
    fun watchNextTransactionSize(): Observable<NextTransactionSize?> {
        return transactionSizePreference.asObservable()
            .map { json: String? ->
                if (json != null) {
                    return@map SerializationUtils.deserializeJson(
                        NextTransactionSize::class.java, json
                    )
                } else {
                    return@map null
                }
            }
    }

    /**
     * Replace the stored NextTransactionSize.
     */
    fun setTransactionSize(transactionSize: NextTransactionSize) {
        transactionSizePreference.setNow(
            SerializationUtils.serializeJson(NextTransactionSize::class.java, transactionSize)
        )
    }

    /**
     * Migration to init expected debt for pre-existing NTSs.
     */
    fun initExpectedDebt() {
        val hasNts = sharedPreferences.contains(KEY_TRANSACTION_SIZE)
        if (!hasNts) {
            return
        }
        val nts = nextTransactionSize
        Preconditions.checkNotNull(nts)
        setTransactionSize(nts!!.initExpectedDebt())
    }

    /**
     * Migration to init outpoints for pre-existing NTSs.
     */
    fun initNtsOutpoints() {
        val hasNts = sharedPreferences.contains(KEY_TRANSACTION_SIZE)
        if (!hasNts) {
            return
        }
        val nts = nextTransactionSize
        Preconditions.checkNotNull(nts)
        setTransactionSize(nts!!.initOutpoints())
    }

    /**
     * Migration to init utxo status for pre-existing NTSs.
     */
    fun initNtsUtxoStatus() {
        val hasNts = sharedPreferences.contains(KEY_TRANSACTION_SIZE)
        if (!hasNts) {
            return
        }
        val nts = nextTransactionSize
        Preconditions.checkNotNull(nts)
        setTransactionSize(nts!!.initUtxoStatus())
    }
}